src/main/java/io/strimzi/kafka/bridge/SourceBridgeEndpoint.java

```diff
@@ -82,22 +81,40 @@ protected void handleClose() {
     /**
      * Send a record to Kafka
      *
-     * @param krecord   Kafka record to send
-     * @param handler   handler to call if producer with unsettled is used
+     * @param record Kafka record to send
+     * @param isAsync if the send doesn't need to get metadata via the Kafka related callback
+     * @return a CompletionStage bringing the metadata
      */
-    protected void send(KafkaProducerRecord<K, V> krecord, Handler<AsyncResult<RecordMetadata>> handler) {
-
-        log.debug("Sending record {}", krecord);
-        if (handler == null) {
-            this.producer.send(krecord);
-        } else {
-            this.producer.send(krecord, handler);
+    protected CompletionStage<RecordMetadata> send(ProducerRecord<K, V> record, boolean isAsync) {
+        CompletableFuture<RecordMetadata> promise = new CompletableFuture<>();
+        Runnable r = () -> {
+            log.debug("Sending record {}", record);
+            log.trace("Send thread {}", Thread.currentThread());
+            if (isAsync) {
+                this.producer.send(record);
+                promise.complete(null);
+            } else {
+                this.producer.send(record, (metadata, exception) -> {
+                    log.trace("Callback thread {}", Thread.currentThread());
+                    if (exception == null) {
+                        promise.complete(metadata);
+                    } else {
+                        promise.completeExceptionally(exception);
+                    }
+                });
+            }
+        };
```

Comment:
Yeah good point, I will do.

Reply:
I answered too quickly :-) ... Going through this I don't see any good solution to get what you are asking for, or I didn't understand what you mean or just ... I don't know :-)
What I see I can do is about extracting a private method like `private void internalSend(ProducerRecord<K, V> record, boolean isAsync, CompletableFuture<RecordMetadata> promise)` and then having:

```java
Runnable r = () -> internalSend(record, isAsync, promise);
```

Anyway it doesn't solve what you said about `creating the Runnable in every call`.
The code in the lamba has a kind of "status" catching the parameters so the runnable is different every call.

What did I get wrong about what you meant here?

Reply:
(In any case if it is not possible it is not possible of course)

Reply:
Well I showed above that the code in the lambda depends on record, isAsync and the promise variables.
The Runnable is needed because provided to the 'runAsync` in order to execute it asynchronously but we know that a Runnable cannot have input parameters, it has just the `run()` method.

Reply:
Yeah, you are right. Makes sense.